Initial implementation of persisted queries plugin

santino opened the pull request:
## Description

This draft PR is for implementing a persisted queries plugin with the following goals:
- Handle multiple persisted queries list, to potentially support multiple clients integrating with a single GraphQL server
- Provide a basic implementation for a store that handle static JSON files (as generated by most common clients)
- Provide a mechanism to update the store at runtime, even targeting specific lists only (whitelist)

Since the plugin is meant to support multiple JSON files, it also offers an option to match from a single list only.
This allows developers to implement logic to target a single list, so you can avoid matching against all lists when you know your request comes from a specific application and so must be available in a single list.
The point is not necessarily to improve performance, in fact matching against multiple lists might still be implemented with o(1) cost (since it's just a lookup job), but more about offering proper support for multiple applications; in fact it might be that different applications have one or more equal queries, that would result in the same SHA256 id. By handling multiple and separate queries lists there is no worry of conflicts and file might just be provided to the server as they're generated by clients.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Being just a draft, this is open for a conversation.
I am planning to add tests once we finalise the strategy.

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

This plugin might be seen in conflict with the existing persisted-oeprations, but it differs as it is more intended (but not limited) to handle static lists.  
The main difference is the concept of handling multiple lists to support multiple client applications.  
Also this plugin provides an implementation for a store (again targeting a basic usage with static files), which is useful to be used as a guide for developers wanting to implement their own store from other sources (such as databases).

Ultimately I am open to merge the two plugins so that we offer a solution that is flexible enough and provides a basic store implementation.

santino committed:
Initial implementation of persisted queries plugin

santino committed:
Add request setup to README draft

santino committed:
updating types and adding more details to the README

santino committed:
Do not cache store, to allow refreshing it on the fly

also better methods naming

n1ru4l commented on packages/plugins/persisted-queries/README.md:
````diff
@@ -0,0 +1,63 @@
+## Getting Started
+
+```
+yarn add @envelop/persisted-queries
````
Isn't persisted-operations the better wording?

santino replied:
Currently I am intentionally using this to differentiate from existing "persisted-operations" plugin.
This is currently a draft to discuss how to progress given overlap, if it encounters favour this might be merged within existing "persisted-operations"; it's something to discuss with project owners.

BTW, if you want my opinion, it's arguable.
You send a request to your server, and a single request can potentially contain multiple operations; yet it would have only a single hash/id.
In this case you're not persisting the operation (as you don't have an id per operation), but the whole query string body.
Hence, you are persisting the query string value, which content can involve multiple operations.

Also, persisted queries seems to be the de-facto convention, as documented in [Relay](https://relay.dev/docs/guides/persisted-queries/) and [Apollo](https://www.apollographql.com/docs/apollo-server/performance/apq/), just to mention some big players; but there are definitely more.

Personally, I would prefer to call it "persisted-documents" because in reality the value that you pass into query string is a "to-be parsed" GraphQL document, but I suppose that would be more difficult to guess, given it's not currently standardised like this.
In fact, if you look into the code, I am setting the hash as "documentId" within the context object; for potential later reference.

yaacovCR replied:
https://github.com/graphql/graphql-spec/pull/777

Persisted documents seems the currently correct term.

documents, of course, are written in “graphQL” or even the graphql query language, still a little wierd

santino replied:
Sounds like I am backed, and so I would be happy to do so!
Will check with owners of the project and eventually change the name of this plugin to persisted-documents before fully opening the PR.

I think eventually, in case we decide to favour this over current "persisted-operations" plugin; the existing one could simply be deprecated and removed allowing this to become the standard.

Again, really looking forward to hear from The Guild on how to solve the overlap first.

tvvignesh replied:
@santino Regarding the naming, we use the name "Persisted Operations" in Hive as of today. So, my view is that irrespective of what name we choose, we should try to stick to one name in both the plugin as well as hive to maintain consistency. Otherwise, both are explanatory according to me.

And in context with overlap with [persisted-operations](https://github.com/dotansimha/envelop/tree/main/packages/plugins/persisted-operations) - you are right, as a user it would sound very confusing to have both `persisted-operations` and `persisted-queries`

There is one question I have here when talking about the overlap, what are the differences you see from your perspective between the both plugins?

@kamilkisiela Thoughts?

santino replied:
My opinion is that persisted-documents is semantically correct and persisted-operations isn't, because as explained above you're not persisting operations but full GraphQL documents.
@yaacovCR seems to agree and has provided a good reference for glossary.

Having said this, as the owners of the project you are definitely free to make your decision, I just want to provide a recommendation.
I totally get your point that The Guild should use the same conventions across all their tools, it will surely be appreciated by your users :)

As about the question about the difference, I have added a separate comment to keep the conversations separate.

dotansimha replied:
That's a endless discussion :D 
We used the term `documents` in codegen, and got some user rejections because they were used to `queries` (which doesn't make sense)... 

Basically, for me, it's:
* `document` - a DocumentNode, that can hold any kind of AST.
* `operation` - Anything executable (query/mutation/subscription).

santino replied:
And there is also Source, to throw more stuff into the mix.
Technically the string containing the Document, which includes the Operations, is a GraphQL Source.

Anyway, was a nice discussion and I care a lot about semantics, but I don't feel like insisting.
Persisted-operations be it, if you feel like so :)

santino committed:
return Promise from load() method, to allow custom error handling

also renamed build() to load()

santino committed:
implement whitelist functionality

useful to update existing store only on specific lists rather than the full array

dotansimha commented on packages/plugins/persisted-queries/src/plugin.ts:
```diff
@@ -0,0 +1,96 @@
+  [contextProperty]?: string;
+}
+
+export type UsePersistedQueriesOptions = {
```
I wonder if these can be moved into the existing `persisted-operations` plugin? 
And then `pickSingleList` might be flag for the store?

santino replied:
Yes, but then you need to be prescriptive of store shape (key/value Map).
So your signature changes.

dotansimha commented on packages/plugins/persisted-queries/src/plugin.ts:
```diff
@@ -0,0 +1,96 @@
+  return {
+    onParse({ context, params, extendContext, setParsedDocument }) {
+      const store = options.store.get(); // retrieve fresh instance of store Map
+      const queryId = options.setQueryId ? options.setQueryId(context) : queryIdFromSource(params.source);
```
I think we have a similar thing in persisted-operations plugin, that @kamilkisiela added

santino replied:
I think currently existing persisted-operations doesn't receive context, to begin with.
Also there is no logic to grab the id/hash from standard GraphQL source.

Probably the bit you're referring to has been introduced by @kamilkisiela in Hive rather then Envelop

dotansimha reviewed (commented):
I like this concept. I wonder if some of the things can be merged with persisted-operations plugin? 🤔
